kip-0015.md

```diff
@@ -0,0 +1,304 @@
+
+# Specification
+
+## Schema Sketch
+
+### Request
+```
+{
+  "cmds": [ { "cmd": string, "sigs": [ { "pubKey":string, "sig": string|null }]}]
+}
+```
+
+### Response
+
+```
+{
+  "responses": [{
+    "commandSigData": { "cmd": string, "sigs": [ { "pubKey":string, "sig": string|null }]},
+    "outcome": { "result": "success|failure|noSig", "hash|message"|null: string } }
+  }]
+}
+```
+or
+```
+{
+  "error": {  "type": "reject|emptyList|other", "msg": string|null }
+}
+```
+
+## **Quicksign Request**
+
+A quicksign request takes the form of a json-encoded payload matching the following schema:
+
+```
+{
+  "cmds": [CommandSigData]
+}
+```
+
+## CommandSigData
+
+- The "cmd" string represents a json-encoded payload type, specified here: [https://api.chainweb.com/openapi/pact.html#tag/model-payload](https://api.chainweb.com/openapi/pact.html#tag/model-payload)
+    - Here is an example `cmd` payload:
+
+```
+    "{\"networkId\":\"testnet04\",\"payload\":{\"exec\":{\"data\":null,\"code\":\"(+ 1 2)\"}},\"signers\":[{\"pubKey\":\"ae18efd16cbd49e9a92552a2589ac01491b486fbcbb1e3f07980f945597e2033\",\"clist\":[{\"args\":[],\"name\":\"coin.GAS\"}]},{\"pubKey\":\"fa781bdd858cd2380b5e2b654e58035f7189a6e8158686a1bb7eabb585a56e7f\"}],\"meta\":{\"creationTime\":1663085494,\"ttl\":1800,\"gasLimit\":20,\"chainId\":\"0\",\"gasPrice\":1.0e-6,\"sender\":\"ae18efd16cbd49e9a92552a2589ac01491b486fbcbb1e3f07980f945597e2033\"},\"nonce\":\"2022-09-13 16:11:34.678231 UTC\"}"
+```
+
+The schema for a `CommandSigData` object is:
+
+```
+{
+  "sigs": [Signer],
+  "cmd": string
+}
+```
+
+Note that this type is an evolved version of `SigData` which pact currently uses in its command-line tool: [https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66](https://github.com/kadena-io/pact/blob/957b8bd7644cebc60d043b33f0d7ffc53c65d783/src/Pact/Types/SigData.hs#L62-L66)
+
+## Signer
+
+- The `pubKey` field is a lowercase base-16 encoded public key
+- The `sig` element is either a lowercase base-16 encoded signature or a null value (representing a request for signature).
+
+```
+{
+  "pubKey": string,
+  "sig": string //optional; its omission represents a signature that has not yet been acquired
+}
+```
+
+## **Quicksign Response**
+```
+{
+  "responses": [Response]
+}
+```
+
+or
+
+```
+{
+  "error": QuicksignError
+}
+```
+
+### Response
+
+- The `cmds` field contains the list of requests; some of which contain new signatures attached. Even in the case of an error, it is still useful to have the ordered list of `CommandSigDatas` in the response, given that some of them will contain new signatures.
+
+```
+{
+  "commandSigData": CommandSigData,
+  "outcome": Outcome
+}
+```
+
```

Comment:
CommandSigData from the request, is not the same as it is from the response. Fields are not optional in the response that are in the request.

- `sigs` will not contain nulls, as there will always be a `{ "pubKey": string }`
- so `Signer`-type will have `"pubKey": string`
- `sig` will be `null` to make explicit that there's a signature missing

```suggestion
## CommandSigDataResponse

```ts
type CommandSigDataResponse = {
  "sigs": [SignerResponse],
  "cmd": string
}

type SignerResponse = {
  "pubKey": string,
  "sig": string | null | undefined,
}
```

```